Make ServerConn safe for concurrent use

brianfoshee opened the pull request:
ServerConn's textproto.Conn is safe for concurrent access. This adds
calls to StartResponse() and EndResponse() to put that to use wherever
Cmd() is called.
https://golang.org/pkg/net/textproto/#Conn.Cmd

Additionally, this adds a sync.Mutex on ServerConn to provide
concurrency protection for data commands such as STOR and RETR that need
to run through synchronous steps.

jlaffaye commented:
I dont understand why cmdDataConnFrom() needs to return the id.
We can call EndResponse right after
`code, msg, err := c.conn.ReadResponse(-1)` line 176

brianfoshee commented:
The id is necessary only when being used in `StorFrom()`. That method calls 
`_, _, err = c.conn.ReadResponse(StatusClosingDataConnection)` which needs to be wrapped by `StartResponse`/`EndResponse`.

Another way to go about it, to avoid the method signature change, would be to add a private method that is a copy of the `cmdDataConnFrom()` in this commit, change `cmdDataConnFrom()` back to its old signature, have `cmdDataConnFrom()` call that new method and `End` the response but return everything else, then use the new private method inside of `StorFrom()`.

brianfoshee commented:
I pushed a branch with that option. I haven't had a chance to test it out yet though. https://github.com/brianfoshee/ftp/commit/88fadc2bc9699eb4068172b8ddf9cead28e4e5bd

brianfoshee commented:
Update: tests don't pass when using that option. The response needs to be held throughout the entire command.

I should probably change my wording on the PR. This will not enable concurrent uploads, per se, but it will make a single ServerConn instance safe to be passed around to multiple goroutines which is the use-case that I'm looking for. The StartResponse/EndResponse calls are safety checks to avoid contention.

brianfoshee commented:
I've found an issue with EndResponse() being called too early on the commands other than STOR. I'm reworking the PR to account for it, including a new test.

brianfoshee commented:
@jlaffaye I've updated this PR to support concurrent access to a shared `ServerConn`, including a test to verify that behavior. A `sync.Mutex` is required around all of the data actions (anything that calls `cmdDataConnFrom`) because those require a synchronous sequence of events. I also removed the change to `cmdDataConnFrom()` and instead store the pipeline `id` for all data actions on the `ServerConn` itself. Note the change to `response.Close()` - the mutex is Unlocked and the Pipeline response is Ended.

corny commented on ftp.go:
```diff
@@ -238,17 +247,29 @@ func (c *ServerConn) openDataConn() (net.Conn, error) {
+	// sequence of events. Must be Unlock()'d before any other data action can
+	// proceed.
+	c.mu.Lock()
+
```
Why don't you put a `defer c.mu.Unlock()` here?
There are returns without unlocking before.

brianfoshee replied:
The corresponding `c.mu.Unlock()` is dependent on the calling method. For instance, the `StorFrom()` method reads the data it needs from the connection, closes the connection, and is done with that transaction, so to speak. So that method is okay to `defer c.mu.Unlock()`. However, in the case of `RetrFrom()`, where it returns a `response` struct as its `io.ReadCloser`, the `c.mu.Unlock()` happens when `Close()` is called on the `response` since that is closed by code outside of this package.

It comes down to needing to lock the entire transaction around the `net.Conn` being opened, any data being read or written to it, and it being closed. Otherwise, deadlock occurs.

corny replied:
Oh sorry, i didn't see the comment above the method.

corny replied:
I reviewed the code again. If this functions returns an error there are cases in which unlocking is done before and in othere case the unlocking is not done.

I would like to see more closures to avoid these kind of mistakes.

corny commented:
I created a branch with closures for data connections: https://github.com/jlaffaye/ftp/compare/master...digineo:data-closures

Still untested, but locks for concurrency could be added easily to this version.
